client/asset/btc/btc.go

```diff
@@ -660,6 +660,79 @@ func (btc *ExchangeWallet) feeRateWithFallback(confTarget uint64) uint64 {
 	return feeRate
 }
 
+// MaxOrder generates information about the maximum order size and associated
+// fees that the wallet can support for the given DEX configuration. The fees are an
+// estimate based on current network conditions, and will be <= the fees
+// associated with nfo.MaxFeeRate. For quote assets, the caller will have to
+// calculate lotSize based on a rate conversion from the base asset's lot size.
+func (btc *ExchangeWallet) MaxOrder(lotSize uint64, nfo *dex.Asset) (*asset.OrderEstimate, error) {
+	networkFeeRate, err := btc.feeRate(1)
+	if err != nil {
+		return nil, fmt.Errorf("error getting network fee estimate: %w", err)
+	}
+	btc.fundingMtx.RLock()
+	utxos, _, avail, err := btc.spendableUTXOs(0)
+	btc.fundingMtx.RUnlock()
+	if err != nil {
+		return nil, fmt.Errorf("error parsing unspent outputs: %w", err)
+	}
+	// Start by attempting max lots with no fees.
+	lots := avail / lotSize
+	for lots > 0 {
+		val := lots * lotSize
+		sum, size, _, _, _, _, err := btc.fund(val, lots, utxos, nfo)
+		// The only failure mode of btc.fund is when there is not enough funds,
+		// so if an error is encountered, count down the lots and repeat until
+		// we have enough.
+		if err != nil {
+			lots--
+			continue
+		}
+		reqFunds := calc.RequiredOrderFunds(val, uint64(size), lots, nfo)
+		maxFees := reqFunds - val
+		estFunds := calc.RequiredOrderFundsAlt(val, uint64(size), lots, nfo.SwapSizeBase, nfo.SwapSize, networkFeeRate)
+		estFees := estFunds - val
+		// Math for split transactions is a little different.
+		if btc.useSplitTx {
+			_, extraFees := btc.splitBaggageFees(nfo.MaxFeeRate)
+			_, extraEstFees := btc.splitBaggageFees(networkFeeRate)
+			if avail >= reqFunds+extraFees {
+				return &asset.OrderEstimate{
+					Lots:          lots,
+					Value:         val,
+					MaxFees:       maxFees + extraFees,
+					EstimatedFees: estFees + extraEstFees,
+					Locked:        val + maxFees + extraFees,
+				}, nil
+			}
+		}
```

Comment:
What if `avail < reqFunds+extraFees`?  It should return `&asset.OrderEstimate{}` right?  Instead it seems to go with the no split txn estimate.

Reply:
The estimate is like a dry run. Falling back to non-split is what will happen when they order.

Reply:
OK, I honestly did not recall that it falls back to non-split if needed.  That actually sounds undesirable as it violates the user's settings.

Reply:
It only falls back when the excess is not enough to cover the fees associated with the split. In that case, we are doing them a favor by not using a split transaction. Overall cost is lower. Split only makes sense when the excess is substantial.

Reply:
Ah, I was aware of that.  Dunno how I got that mixed up.  All good.